### PR TITLE
build(RND-7271): MIgrate to confluent-kafka lib

### PR DESCRIPTION
**Pull request by Christos-Hadjinikolis**
## Overview

This PR addresses the need to ensure compatibility with Kafka brokers version 3.5. During our review, we discovered that our current Kafka Python library was outdated (not updated since 2020) and was not the official Confluent-Kafka library, which is better maintained and more up-to-date. Therefore, we decided to migrate from Kafka-Python to Confluent-Kafka. This PR implements the necessary changes for this migration.

## Key Changes

* Migrated from Kafka-Python to Confluent-Kafka in all relevant modules.

## Impact

* Ensures compatibility with Kafka brokers version 3.5.
* Improves the maintainability and reliability of our Kafka interactions by using the official Confluent-Kafka library.
* **Removing support for Kafka-Python --> breaking change!**

## Checklist

The following won't apply in all cases - mark `N/A` next to point if not needed.

- [x] Appropriate unit tests added/updated
- [x] Module, function, class docstrings updated
- [x] Comments added to PR where necessary
- [x] Interface documentation updated
- [x] Semantic commits used for this PR, so that a CHANGELOG can be generated from them (don't delete your local branch! when tagging a new release from master)

## Release-Steps

- [x] Merge PR
- [x] Release new tag

### CODE REVIEW

File: dynamicio/mixins/with_kafka.py [hunk 1]
```diff
@@ -124,39 +297,39 @@ def _get_producer(self, server: str, **options: MutableMapping[str, Any]) -> Kaf
...
+        self.__producer.flush()
 
     @staticmethod
-    def _default_key_serializer(key: Optional[str]) -> Optional[bytes]:
-        if key:
-            return key.encode("utf-8")
+    def _on_delivery(err, msg):
+        """Callback for message delivery."""
+        if err is not None:
+            raise Exception(f"Message delivery failed: {err}, for message: {msg}")
```
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins/with_kafka.py:325]: Single message failing to be sent will trow an exception as demonstrated below: 
```shell
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Traceback (most recent call last):\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/.pyenv/versions/3.10.4/lib/python3.10/multiprocessing/pool.py", line 125, in worker\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    result = (True, func(*args, **kwds))\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 130, in send_voyage_documents\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    raise Exception(error_msg) from generic_exception\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Exception: Exception: Message delivery failed: KafkaError{code=NOT_ENOUGH_REPLICAS,val=19,str="Broker: Not enough in-sync replicas"}, for message: <cimpl.Message object at 0x7fc2ed58ae40>\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Traceback (most recent call last):\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 82, in send_voyage_documents\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    kafka_io_write(wrapped_documents, mode)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 171, in kafka_io_write\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    _connect_with_retry(lambda: kafka_io.write(pd.DataFrame.from_records(documents_as_dicts).set_index("key")), _KAFKA_CONNECTION_ATTEMPTS, _KAFKA_CONNECTION_BACKOFF_SECS)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 136, in _connect_with_retry\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    return write_op()\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/src/msg_producer/main.py", line 171, in <lambda>\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    _connect_with_retry(lambda: kafka_io.write(pd.DataFrame.from_records(documents_as_dicts).set_index("key")), _KAFKA_CONNECTION_ATTEMPTS, _KAFKA_CONNECTION_BACKOFF_SECS)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/core.py", line 185, in write\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    getattr(self, f"_write_to_{source_name}")(self._apply_schema(df))\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/mixins/with_kafka.py", line 259, in _write_to_kafka\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    self._send_messages(df=df, topic=self.sources_config.kafka.kafka_topic)\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/mixins/with_kafka.py", line 319, in _send_messages\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    self.__producer.flush()\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'  File "/home/circleci/project/dynamicio/mixins/with_kafka.py", line 325, in _on_delivery\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'    raise Exception(f"Message delivery failed: {err}, for message: {msg}")\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'Exception: Message delivery failed: KafkaError{code=NOT_ENOUGH_REPLICAS,val=19,str="Broker: Not enough in-sync replicas"}, for message: <cimpl.Message object at 0x7fc2ed58ae40>\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'\n'
[2024-05-28 09:57:39,646] {pod_launcher.py:166} INFO - b'"""\n'
```

There is room to relax this condition with a parameter that controls how sensitive the application should be. I.e. throw exceptions after 10-20-100 failed deliveries, else just log them... 

I think the library should be super-sensitive to message sending though, hence why I implemented it like this.

File: dynamicio/mixins/with_kafka.py [hunk 2]
```diff
@@ -76,9 +73,174 @@ class WithKafka:
...
     __kafka_config: Optional[Mapping] = None
-    __producer: Optional[KafkaProducer] = None
+    __producer: Optional[Producer] = None
     __key_generator: Optional[Callable[[Any, Mapping[Any, Any]], Optional[str]]] = None
     __document_transformer: Optional[Callable[[Mapping[Any, Any]], Mapping[Any, Any]]] = None
+    __key_serializer: Optional[Callable[[Optional[str]], Optional[bytes]]] = None
+    __value_serializer: Optional[Callable[[Mapping[Any, Any]], bytes]] = None
+
+    # N.B.: Please refer to https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md and update this config in case of a major release change.
+    VALID_CONFIG_KEYS = {
```
Comment (joloppo) [inline dynamicio/mixins/with_kafka.py:83]: I wouldn't check this stuff. Just pass it onto kafka and it should handle however it handles these.

Comment (joloppo) [reply dynamicio/mixins/with_kafka.py:83]: Nvm I forgot the "put everything into a dictionary then fish out stuff" pattern.

File: dynamicio/mixins/with_kafka.py [hunk 3]
```diff
@@ -89,28 +251,39 @@ def _write_to_kafka(self, df: pd.DataFrame) -> None:
...
                 self.__document_transformer = self.options.pop("document_transformer")
-
-        if self.__producer is None:
-            self.__producer = self._get_producer(self.sources_config.kafka.kafka_server, **self.options)
-
-        self._send_messages(df=df, topic=self.sources_config.kafka.kafka_topic)
-
-    @utils.allow_options(KafkaProducer.DEFAULT_CONFIG.keys())
-    def _get_producer(self, server: str, **options: MutableMapping[str, Any]) -> KafkaProducer:
+        if self.__key_serializer is None:
```
Comment (joloppo) [inline dynamicio/mixins/with_kafka.py:271]: You can just set __key_serializer and __value_serializer to whatever your default functions are, then override if options contain either. Not sure why this indirection exists.

**Review by joloppo — APPROVED** [review]
If it runs. Those tests look very dicey. I put one comment.

File: dynamicio/mixins/with_kafka.py (hunk 2 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/mixins/with_kafka.py:83]: yes... one of the things we should change for sure once we migrate away from the mixins pattern

File: tests/mocking/io.py [hunk 4]
```diff
@@ -117,11 +119,15 @@ class WriteKeyedKafkaIO(UnifiedIO):
...
 
 class MockKafkaProducer:
-    def __init__(self):
+    def __init__(self, **kwargs):
         self.my_stream = []
+        self.config = kwargs  # Store the config to ensure it can accept any options
+        self.produce_call_count = 0
 
-    def send(self, topic: str, value: dict, key: str = None):  # pylint: disable=unused-argument
+    def produce(self, topic: str, key: Optional[bytes], value: bytes, on_delivery: Callable):  # pylint: disable=unused-argument
```
Comment (deaglancrew) [inline tests/mocking/io.py:127]: Is there a reason you're mocking out the kafka calls? Can we not have an actual kafka deployed alongside in CI?

Comment (Christos-Hadjinikolis) [reply tests/mocking/io.py:127]: Great idea... though needs more effort. happy to implement in a later PR!

File: dynamicio/mixins/with_kafka.py (hunk 3 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/mixins/with_kafka.py:271]: Looking into it...

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -58,17 +58,6 @@ repos:
         types: [file, yaml]
         stages: [commit]
 
-  - repo: local
-    hooks:
-      - id: mypy
```
Comment (deaglancrew) [inline .pre-commit-config.yaml:63]: Is there a reason this has been removed in this PR?

Comment (Christos-Hadjinikolis) [reply .pre-commit-config.yaml:63]: It was never part of the CI anyway and it was not respected. Currently, there are multiple mypy errors that emerge every time the hookds get activated and it is a lot of work to go around and fix em.

File: dynamicio/mixins/with_kafka.py (hunk 3 above)
Comment (Christos-Hadjinikolis) [reply dynamicio/mixins/with_kafka.py:271]: Good point but leaving it as is for now as the intended behaviour is properly tested and the refactoring wont change that behaviour. Good point tho.

**Pull request merged**